Restaurants list

omer-katzir opened the pull request:
**Viewing the restaurants list in the client**

notes:
UI - using Material-UI
we_eat server json api with async 'fetch'
react-redux integration: actions and containers
temporary: added "**Refresh**" button for debugging trigger

omer-katzir committed:
restaurant list ui,fetch,redux

omer-katzir committed:
some corrections and trimming

kobim commented on .gitignore:
```diff
@@ -18,6 +18,12 @@
 .env.production.local
 
+#deprecated
+deprecated/*
```
`depreacted`? on a new project?

SergeyPanchenkoWeWork replied:
why is there a deprecated folder?

kobim commented on .idea/workspace.xml:
```diff
@@ -2,14 +2,13 @@
 <project version="4">
```
you need to `git rm --cached` this file

kobim commented on package-lock.json:
```diff
@@ -0,0 +1,13118 @@
+{
```
we use [yarn](https://yarnpkg.com/en/), not npm.
Please make sure to switch between the two (should be almost seamless)

kobim commented on package.json:
```diff
@@ -3,9 +3,17 @@
   "dependencies": {
+    "@material-ui/core": "^3.9.3",
+    "@material-ui/icons": "^3.0.2",
+    "material-ui-icons": "^1.0.0-beta.36",
```
isn't this a duplicate of `@material-ui/icons`?

kobim commented on package.json:
```diff
@@ -3,9 +3,17 @@
+    "react-scripts": "3.0.1",
+    "redux": "^4.0.1",
+    "redux-thunk": "^2.3.0",
+    "typeface-roboto": "0.0.54"
```
is this being used? I didn't see any instance of usage

kobim commented on src/Actions/Actions.js:
```diff
@@ -0,0 +1,32 @@
+
+export const REQUEST_RESTAURANTS = 'REQUEST_RESTAURANTS';
+export const RECEIVE_RESTAURANTS = 'RECEIVE_RESTAURANTS';
+export const RECEIVED_RESTAURANTS_FAILD = 'RECEIVED_RESTAURANTS_FAILD';
```
typo: missing an `E` - "FAILED"

kobim commented on src/Actions/Actions.js:
```diff
@@ -0,0 +1,32 @@
+}
+
+export function fetchRestaurants() {
+    return (dispatch, getState) => {
```
why do you return a function here?
It doesn't seem like you are using the return value (`Promise`) of `fetch`, so why the need to return a function and not do a simple action?

kobim commented on src/Actions/Actions.js:
```diff
@@ -0,0 +1,32 @@
+export const RECEIVED_RESTAURANTS_FAILD = 'RECEIVED_RESTAURANTS_FAILD';
+
+
+function receivedRestaurantsFailed(e){
```
see [redux-actions](https://github.com/redux-utilities/redux-actions) for an easier approach of writing actions.

kobim commented on src/App.js:
```diff
@@ -1,26 +1,27 @@
-        >
-          Learn React
-        </a>
+        <img src={ require('./restaurant_image.jpg') }
```
pay attention this `require` runs every time `App` renders(), which can be I/O intensive.

kobim commented on src/App.js:
```diff
@@ -1,26 +1,27 @@
+import {createStore, applyMiddleware} from 'redux';
+import rootReducer from './Reducers/Reducers.js'
+
+const store = createStore(rootReducer, applyMiddleware(thunkMiddleware)) ;
```
is `thunk` really required? (see my comment from above)

SergeyPanchenkoWeWork replied:
I would suggest adding https://github.com/reduxjs/redux-devtools - it will make you life easier

kobim commented on src/App.js:
```diff
@@ -1,26 +1,27 @@
 }
 
 export default App;
+//apiRoot='http://localhost:3000/restaurants'/>
```
newline in the end of file

SergeyPanchenkoWeWork replied:
safe to remove

kobim commented on src/Components/Restaurant.js:
```diff
@@ -0,0 +1,75 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { withStyles } from "@material-ui/core/styles/index"
```
minor: you don't have to include `/index`, it's implicit

kobim replied:
please choose either double-quote or single-quote, you don't need both.
Or maybe consider using [prettier](https://github.com/prettier/prettier) or [xo](https://github.com/xojs/xo) to automate the beautification of your code

kobim commented on src/Components/Restaurant.js:
```diff
@@ -0,0 +1,75 @@
+import CardContent from "@material-ui/core/CardContent/index";
+import Typography from "@material-ui/core/Typography/index";
+import Divider from "@material-ui/core/Divider/index";
+import Rating from 'material-ui-rating';
```
cheater ;)

omer-katzir replied:
^-^ חחחח

kobim commented on src/Components/Restaurant.js:
```diff
@@ -0,0 +1,75 @@
+    {
+        tooltip: {
+            color: 0xFF0000,
+            //padding: '4px 8px',
```
remove unused code

kobim commented on src/Components/Restaurant.js:
```diff
@@ -0,0 +1,75 @@
+    }).isRequired,
+};
+
+export default Restaurant;
```
newline at the end of file

kobim commented on src/Components/Restaurant.js:
```diff
@@ -0,0 +1,75 @@
+);
+
+Restaurant.propTypes = {
+    restaurant: PropTypes.shape({
```
Why does a `<Restaurant>` receive an instance of `restaurant` rather the props themselves?
I would want to use this component in this way:
```jsx
<Restaurant id="1234" name="Best Burger" />
```

This will also prevent the usage of `PropTypes.shape`

omer-katzir replied:
please check ./components/GridView.js:31, this allow passing the instance:

`                        <Restaurant restaurant={restaurant}/>`

for individual props the usage can be:
`                <Restaurant restaurant={{id:"awa",cuisine:'american',name:'name',rating:2.5 }}/>
    `

kobim commented on src/Components/Restaurant.js:
```diff
@@ -0,0 +1,75 @@
+    <CardCustom  className={restaurant.name + 'Card'} raised={true}>
+        <CardContent>
+            <Typography gutterBottom variant='h5' component='h2' style={{'color':'#86cfff'}}>
+                                <span style={{'fontFamily':'CuisinesFont', 'fontSize':'150%', 'color':'#c06931'}}>
```
why aren't you using material-ui's `withStyles` here too?

SergeyPanchenkoWeWork replied:
Better use css for it

kobim commented on src/Reducers/Reducers.js:
```diff
@@ -0,0 +1,33 @@
+import {REQUEST_RESTAURANTS} from "../Actions/Actions";
```
import all the constants at once, no need for 3 lines

kobim commented on src/Reducers/Reducers.js:
```diff
@@ -0,0 +1,33 @@
+    action) {
+    switch(action.type){
+        case REQUEST_RESTAURANTS:
+            return Object.assign({}, state = {
```
Use the [spread operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax) here too (Object.assign is 👎 )

kobim replied:
what is `state = {` ?

SergeyPanchenkoWeWork replied:
you can use spread instead of `Object.assign`
```javascript
 return {
  ...state,
 isFetching: true, 
};
```

kobim commented on src/Reducers/Reducers.js:
```diff
@@ -0,0 +1,33 @@
+    switch(action.type){
+        case REQUEST_RESTAURANTS:
+            return Object.assign({}, state = {
+                isFetching: true,
```
what about cleaning the previous error?
maybe cleaning the items too?

omer-katzir replied:
the cleanup is done in the RestaurantsList::mapStateToProps 
_function mapStateToProps(state) {
    const restaurants = state.items || [];
    const isFetching = state.isFetching ||  false;
    const lastUpdated = state.receivedAt;
    const error = state.error;_

kobim commented on src/Reducers/Reducers.js:
```diff
@@ -0,0 +1,33 @@
+    }
+}
+
+export default restaurants;
```
newline

kobim commented on src/Reducers/Reducers.js:
```diff
@@ -0,0 +1,33 @@
+            return Object.assign({}, state = {
+                isFetching: false,
+                items: action.items,
+                lastUpdate: action.receivedAt
```
any reason for having both `lastUpdate` and `receivedAt`? both should be the same.

kobim commented on src/Reducers/Reducers.js:
```diff
@@ -0,0 +1,33 @@
+        case RECEIVE_RESTAURANTS:
+            return Object.assign({}, state = {
+                isFetching: false,
+                items: action.items,
```
what about cleaning the previous error (if any)?

omer-katzir replied:
see previous comment, cleanup is done in the RestaurantsList::mapStateToProps

kobim commented on src/Components/Restaurants.js:
```diff
@@ -0,0 +1,28 @@
+import React, { Component } from 'react';
```
is this file in use? I don't think so.

SergeyPanchenkoWeWork replied:
I don't think this being used currently

kobim commented on src/Containers/RestaurantsList.js:
```diff
@@ -0,0 +1,79 @@
+
+function mapStateToProps(state) {
+    const restaurants = state.items || [];
+    const isFetching = state.isFetching ||  false;
```
why isn't there a default for `isFetching`?
also, pay attention that if `isFetching = false`, this "OR" condition will still execute (and evaluate to `false`)

kobim commented on src/Containers/RestaurantsList.js:
```diff
@@ -0,0 +1,79 @@
+
+const mapDispatchToProps = (dispatch, ownProps) => {
+    return {
+        getRestaurants: () => dispatch(fetchRestaurants()),
```
see https://dev.to/mcrowder65/you-ve-been-doing-mapdispatchtoprops-wrong-this-entire-time-582m

omer-katzir replied:
need your help with that, how do I pass the dispatch function forward to fetchRestaurants,
it has async usage of dispatch in it's promise

kobim commented on src/Containers/RestaurantsList.js:
```diff
@@ -0,0 +1,79 @@
+                        </span>
+                    ) }
+                    { !isFetching && (
+                        <button onClick={() => this.props.getRestaurants()}>Refresh</button>
```
is this material-ui button?

kobim commented on src/Containers/RestaurantsList.js:
```diff
@@ -0,0 +1,79 @@
+                <p>
+                    { lastUpdated && (
+                        <span>
+                            Last updated at {new Date(lastUpdated).toLocaleTimeString()}.{' '}
```
why do you need to `new Date`?

omer-katzir replied:
doesn't work without creating a date object from the timestamp 'lastUpdate'

kobim commented on src/Containers/RestaurantsList.js:
```diff
@@ -0,0 +1,79 @@
+};
+
+function mapStateToProps(state) {
+    const restaurants = state.items || [];
```
why isn't there a default for `items`?

kobim commented on src/Containers/RestaurantsList.js:
```diff
@@ -0,0 +1,79 @@
+    getRestaurants: PropTypes.func.isRequired,
+};
+
+function mapStateToProps(state) {
```
why isn't this function written in arrow notation like to others?

kobim commented on src/Containers/RestaurantsList.js:
```diff
@@ -0,0 +1,79 @@
+    }
+}
+
+const mapDispatchToProps = (dispatch, ownProps) => {
```
you don't need `ownProps`

omer-katzir replied:
use it now for isFetching check

SergeyPanchenkoWeWork commented on .gitignore:
```diff
@@ -18,6 +18,12 @@
+deprecated/*
+
+# ignore local settings
+.idea/*
```
I still see files from .idea in the commit

kobim commented on src/Components/Restaurant.js:
```diff
@@ -0,0 +1,75 @@
+            {restaurant.accept10bis ? 'accepts 10bis' : ''}
+        </div>
+    }>
+    <CardCustom  className={restaurant.name + 'Card'} raised={true}>
```
use [template literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) instead of string concatenation

omer-katzir replied:
removed

kobim commented on src/Components/Restaurant.js:
```diff
@@ -0,0 +1,75 @@
+            id: {restaurant.id} <br/>
+            name: {restaurant.name} <br/>
+            rating: {restaurant.rating.toFixed(1)}<br/>
+            {restaurant.address ? ('address: ' + restaurant.address) : ''}
```
what about a newline after the address?

SergeyPanchenkoWeWork replied:
you can use ` `address: ${restaurant.address}``

kobim commented on src/Components/Restaurants.js:
```diff
@@ -0,0 +1,28 @@
+            <List>
+                {restaurants.map(restaurant =>
+                    <ListItem key={restaurant.id} alignItems='flex-start'>
+                        <Restaurant restaurant={restaurant}/>
```
why `Restaurant` itself isn't a `ListItem`? is it reusable enough?

omer-katzir replied:
should be reusable, just changed the UI to GridList and it still works

omer-katzir replied:
I changed the list to GridList with GridListItem, so need the robustness

kobim commented on src/Components/Restaurants.js:
```diff
@@ -0,0 +1,28 @@
+class Restaurants extends Component {
+    render() {
+        const restaurants = this.props.restaurants;
+        console.log(restaurants);
```
remove

SergeyPanchenkoWeWork commented on package.json:
```diff
@@ -3,9 +3,17 @@
     "react-dom": "^16.8.6",
-    "react-scripts": "3.0.1"
+    "react-redux": "^7.0.3",
+    "react-scripts": "3.0.1",
```
Probably should be in dev

kobim commented on src/Components/Restaurants.js:
```diff
@@ -0,0 +1,28 @@
+import Restaurant from './Restaurant.js';
+
+
+class Restaurants extends Component {
```
If you only have `render()`, use [stateless component](https://hackernoon.com/react-stateless-functional-components-nine-wins-you-might-have-overlooked-997b0d933dbc)

omer-katzir replied:
removed

kobim reviewed (commented):
🍾 first React code!
See comments below, and please use some prettier to make sure the code aligns beautifully.

SergeyPanchenkoWeWork commented on src/Actions/Actions.js:
```diff
@@ -0,0 +1,32 @@
+}
+function receiveRestaurants(json) {
+    return {
+        type: RECEIVE_RESTAURANTS,
```
I would suggest taking a look at https://github.com/redux-utilities/flux-standard-action a standard for writing actions.
and an implementation https://github.com/redux-utilities/redux-actions

SergeyPanchenkoWeWork commented on src/Actions/Actions.js:
```diff
@@ -0,0 +1,32 @@
+
+export function fetchRestaurants() {
+    return (dispatch, getState) => {
+        if (getState().isFetching)
```
Add `{` `}` for more readability

SergeyPanchenkoWeWork commented on src/App.js:
```diff
@@ -1,26 +1,27 @@
 function App() {
   return (
+      <Provider store={store}>
     <div className="App">
```
```javascript
import style from './App.css';

function App() {
  return (
   <div className={style.App}>
...
</div>  
);
}

SergeyPanchenkoWeWork commented on src/Components/Restaurants.js:
```diff
@@ -0,0 +1,28 @@
+class Restaurants extends Component {
+    render() {
+        const restaurants = this.props.restaurants;
+        console.log(restaurants);
```
Maybe you should add `no-console` rule to eslint just not to miss next time

omer-katzir replied:
removed file

SergeyPanchenkoWeWork commented on src/Containers/RestaurantsList.js:
```diff
@@ -0,0 +1,79 @@
+                <p>
+                    { lastUpdated && (
+                        <span>
+                            Last updated at {new Date(lastUpdated).toLocaleTimeString()}.{' '}
```
why `{' '}`

SergeyPanchenkoWeWork commented on src/Containers/RestaurantsList.js:
```diff
@@ -0,0 +1,79 @@
+    }
+}
+
+const mapDispatchToProps = (dispatch, ownProps) => {
```
you can use the shorthand
```javascript
const mapDispatchToProps = {
   getRestaurants: fetchRestaurants
};
```

SergeyPanchenkoWeWork commented on src/Components/Restaurant.js:
```diff
@@ -0,0 +1,75 @@
+import Rating from 'material-ui-rating';
+import Tooltip from '@material-ui/core/Tooltip/index';
+
+const CuisineIconCode = {'american': 65, 'asian': 46, 'bakery': 80,
```
Usually, use all capital for constants

SergeyPanchenkoWeWork commented on src/Components/Restaurant.js:
```diff
@@ -0,0 +1,75 @@
+            {restaurant.accept10bis ? 'accepts 10bis' : ''}
+        </div>
+    }>
+    <CardCustom  className={restaurant.name + 'Card'} raised={true}>
```
why do you need dynamic className that doesn't add any styles?

SergeyPanchenkoWeWork commented on src/Components/Restaurant.js:
```diff
@@ -0,0 +1,75 @@
+        <CardContent>
+            <Typography gutterBottom variant='h5' component='h2' style={{'color':'#86cfff'}}>
+                                <span style={{'fontFamily':'CuisinesFont', 'fontSize':'150%', 'color':'#c06931'}}>
+                                    {String.fromCharCode(CuisineIconCode[restaurant.cuisine.toString().toLowerCase()])} </span>{restaurant.name}
```
Didn't quite get the `fromCharCode` it will return a single char like 'A' for 65

omer-katzir replied:
using it to convert from decimal to unicode char string representation

SergeyPanchenkoWeWork reviewed (commented):
I would suggest trying to keep tabs as 2 and directories as lowerCase

omer-katzir committed:
reducer, actions corrections, idea untrack

omer-katzir committed:
grid view, yarn instead of npm, typos

omer-katzir committed:
thunk remove

omer-katzir committed:
styling corrections

omer-katzir committed:
redux-actions, createActions, handleActions

omer-katzir committed:
restaurant styles

omer-katzir committed:
fixed last updated

omer-katzir committed:
styling

omer-katzir committed:
changed folder names, yarn.lock update

omer-katzir committed:
yarn changes, redux dev tools

omer-katzir committed:
styling and UI corrections

omer-katzir committed:
minor corrections

omer-katzir committed:
correct mapdispatch and added thunk

yyaanniivv commented on src/App.js:
```diff
@@ -1,26 +1,38 @@
+const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
+
+const store = createStore(restaurantsReducer, {
+  isFetching: false,
```
So do we still need these defaults?
I think we can drop them, make sure we have an initial state with in the restaurantReducer that should solve issues with "Data/rendering prior to fetching"

yyaanniivv commented on src/App.js:
```diff
@@ -1,26 +1,38 @@
 function App() {
   return (
-    <div className="App">
+      <Provider store={store}>
```
indentation is off.
make sure you have configured `eslint` with some basics configs. ( i would recommend to take the eslintrc file from spacestation-web repo)

omer-katzir replied:
still don't know why the change wasn't in the previous commits....

yyaanniivv replied:
check if you have soft tabs (spaces) or hard tabs (\t) set in you editor.
it might be two spaces and a tab.

yyaanniivv commented on src/App.js:
```diff
@@ -1,26 +1,38 @@
-        >
-          Learn React
-        </a>
+        <img src={ bkgImage }
```
minor/sidenote:
we try to use full meaningful names.
bkgImage -> backgroundImage

in this case it was clear, but that is not always the case.

yyaanniivv commented on src/Styles.css:
```diff
@@ -0,0 +1,8 @@
+.Restaurants-list{
```
extra minor:
no reason for it not to work with capital case -  most of our css selectors are lower case, as is the file name `style.css`

yyaanniivv commented on src/reducers/Reducers.js:
```diff
@@ -0,0 +1,27 @@
+            }
+        )
+    },
+    {
```
not sure but i think this should change to be in the correct nested position

omer-katzir replied:
not sure what you mean....

yyaanniivv replied:
```
objects =[
  {
    a: ...,
  },
    {
      b: ..,
    },
]
```
Sibling keys should be in the same indentation level.
And in general the code should be consistent as it vastly helps the readability.
Make sure to set up eslint.

yyaanniivv commented on src/reducers/Reducers.js:
```diff
@@ -0,0 +1,27 @@
+
+export const restaurantsReducer = handleActions(
+    {
+        [requestRestaurants]: (state) => (
```
align the indentations here pleaseeeee (eslint!)

yyaanniivv commented on src/containers/RestaurantsList.js:
```diff
@@ -0,0 +1,61 @@
+        const { restaurants, isFetching, lastUpdate, error } = this.props;
+        return(
+            <div className='Restaurants-list'>
+                <p>
```
you can use [React.fragment](https://reactjs.org/docs/fragments.html) if you used the <p> just to bound  these components together.
```
<>
  isFetching
  error
  lastUpdate
</>
```

omer-katzir replied:
if you meant the <p> so removed it

yyaanniivv commented on src/containers/RestaurantsList.js:
```diff
@@ -0,0 +1,61 @@
+                    {error && (<span>{error.message}</span>)}
+                    {lastUpdate && !error && (<span>Last updated at {new Date(lastUpdate).toLocaleTimeString()}</span>)}
+                </p>
+                <GridView restaurants={restaurants}/>
```
you could have conditional rendering here, since you told me you got issues with the default value.
`restaurants && (<GridView restaurants={restaurants} />)`

omer-katzir replied:
I know, but putting a guard here might obscure the right flow and initialisation, anyways: I took the defaultState from the reducer and used it in the preloadedState of the store so there is no overhead now

yyaanniivv commented on src/components/Restaurant.js:
```diff
@@ -0,0 +1,85 @@
+    restaurant: PropTypes.shape({
+        id: PropTypes.string.isRequired,
+        name: PropTypes.string.isRequired,
+        cuisine: PropTypes.string.isRequired, //oneOf(Cuisine).isRequired,
```
remove comment - or create that sub-type and use it.

yyaanniivv commented on src/components/Restaurant.js:
```diff
@@ -0,0 +1,85 @@
+const CUISINE_ICON_CODE = {'american': 65, 'asian': 46, 'bakery': 80,
+                        'fast_food': 87, 'steak': 51, 'sushi': 73, 'vegetarian': 36};
+
+const styles = theme => ({
```
should be extracted to a stylesheet

omer-katzir replied:
actually this is how material-ui documentation uses their 'classes' feature in order to export 'withStyles'
**export default withStyles(styles)(Restaurant);**

yyaanniivv commented on src/actions/Actions.js:
```diff
@@ -0,0 +1,25 @@
+import { createActions } from 'redux-actions'
+
```
These new lines that can be removed.
Try and keep it consistent with a single breakline when needed.

yyaanniivv commented on src/components/Restaurant.js:
```diff
@@ -0,0 +1,85 @@
+import React from 'react';
```
react files - are normaly denoted as .jsx

Also notice that the defined components should be capitalised(name of component and file name) this is also how react understands if the component is standard or custom/user made.

omer-katzir committed:
'RestaurantsList'  changed to 'RestaurantsContainer', 'GridView' changed to 'RestaurantsList',
createStore with combineReducers and added 'restaurants:' hirarchy to store state

omer-katzir committed:
styling and naming

omer-katzir committed:
styling and naming

SergeyPanchenkoWeWork commented on src/reducers/Reducers.js:
```diff
@@ -0,0 +1,28 @@
+            isFetching: false,
+            error: action.error? action.payload : undefined,
+            items: action.error? [] : action.payload,
+            lastUpdate: Date.now()
```
Reducers must be `predictable` which mean you can't use `Date.now()` inside reducers (it will make state different every time it is called)

you should pass it inside the action

omer-katzir committed:
yarn.lock, package.json linting

omer-katzir committed:
linting corrections and config

omer-katzir committed:
scss styles and linting
